File: 2021/src/day12.jl
module Day12

# To use this, add path of AdventOfCodeUtils.jl to 
# ~/.julia/config.startup.jl, eg the line
# push!(LOAD_PATH, "/users/me/dev/shared/src)
using AdventOfCodeUtils

using DelimitedFiles

# using Memoize
# using Combinatorics

# Puzzle description: https://adventofcode.com/2021/day/12

struct Cave
  id::String
  connections::Vector{Cave}
  is_small::Bool

  function Cave(id)
    new(id, Vector{Cave}(), is_lowercase(id))
  end
end

add_connection(from, to) = push!(from.connections, to)
is_start(c) = c.id == START
is_end(c) = c.id == END

START = "start"
END = "end"
visited = Dict{Cave,Int}()

function traverse(cave, part2 = false)
  was_visited() = cave.is_small && visited[cave] > 0 &&
                  (part2 ? small_cave_twice() : true)

  small_cave_twice() = length(filter(kv -> kv[1].is_small && kv[2] > 1,
    visited)) > 0

  result = 0
  get!(visited, cave, 0)

  is_end(cave) && (return 1)
  was_visited() && (return 0)

  visited[cave] += 1

  result = mapreduce(c -> traverse(c, part2), +, filter(!is_start, cave.connections))

  # Chaining (alternative)
  # result = cave.connections |>
  #          c -> filter(!is_start, c) |>
  #               f -> mapreduce(c -> traverse(c, part2), +, f)

  visited[cave] -= 1

  result
end

function solve(cave_system, part2 = false)
  empty!(visited)
  @show traverse(cave_system[START], part2)
end

function read_input(input_file)
  cave_system = Dict{String,Cave}()
  input = readdlm(input_file, String)
  for line in input
    from_id, to_id = split(line, "-")
    from_cave = get!(cave_system, from_id, Cave(from_id))
    to_cave = get!(cave_system, to_id, Cave(to_id))
    add_connection(from_cave, to_cave)
    add_connection(to_cave, from_cave)
  end
  cave_system
end

function main()
  main_input = read_input("../data/day12.txt")
  test_input1 = read_input("../data/day12-test-1.txt")
  test_input2 = read_input("../data/day12-test-2.txt")
  test_input3 = read_input("../data/day12-test-3.txt")

  @assert solve(test_input1) == 10
  @assert solve(test_input2) == 19
  @assert solve(test_input3) == 226

  @assert solve(test_input1, true) == 36
  @assert solve(test_input2, true) == 103
  @assert solve(test_input3, true) == 3509

  @show solve(main_input) # 4186
  @show solve(main_input, true) # 92111
end

@time main()
end

File: 2015/src/day12.py
import sys
import os
import json

sys.path.append(os.path.relpath("../../shared/python"))

# noqa stops autopep8 from reordering this import
from utils import *  # noqa: E402

# Puzzle description: https://adventofcode.com/2015/day/12

DEBUG = True
print_debug = print if DEBUG else lambda *a, **k: None


def read_input(input_file):
    with open(input_file) as f:
        input = json.load(f)
    return input

def json_values_count(obj, part2=False):
    result = 0
    if isinstance(obj, dict):
        if part2 and 'red' in obj.values():
            return 0
        for key, value in obj.items():
            #  print(f"Key: {key}")
            result += json_values_count(value, part2)
    elif isinstance(obj, list):
        for item in obj:
            result += json_values_count(item, part2)
    else:
        #  print(f"Value: {obj}")
        if isinstance(obj, int):
            result += obj

    return result

def solve(input, part2=False):
    return json_values_count(input, part2)


def main():
    input = read_input("../data/day12.txt")

    assert (res := solve([1,2,3])) == 6, f'Actual: {res}'
    assert (res := solve({"a":2,"b":4})) == 6, f'Actual: {res}'
    assert (res := solve([[[3]]] )) == 3, f'Actual: {res}'
    assert (res := solve({"a":{"b":4},"c":-1})) == 3, f'Actual: {res}'
    assert (res := solve({"a":[-1,1]})) == 0, f'Actual: {res}'
    assert (res := solve([-1,{"a":1}])) == 0, f'Actual: {res}'
    assert (res := solve([])) == 0, f'Actual: {res}'
    assert (res := solve({})) == 0, f'Actual: {res}'

    print(f'Part 1 {solve(input)}')  # 191164

    assert (res := solve([1,2,3], True)) == 6, f'Actual: {res}'
    assert (res := solve([1,{"c":"red","b":2},3], True)) == 4, f'Actual: {res}'
    assert (res := solve({"d":"red","e":[1,2,3,4],"f":5}, True)) == 0, f'Actual: {res}'
    assert (res := solve([1,"red",5], True)) == 6, f'Actual: {res}'

    print(f'Part 2 {solve(input, True)}')  # 87842
    
if __name__ == '__main__':
    main()
